pkg/mads/reconcile/reconciler.go

```diff
@@ -0,0 +1,36 @@
+package reconcile
+
+import (
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_cache "github.com/envoyproxy/go-control-plane/pkg/cache"
+
+	util_xds "github.com/Kong/kuma/pkg/util/xds"
+)
+
+func NewReconciler(hasher envoy_cache.NodeHash, cache util_xds.SnapshotCache,
+	snapshotter Snapshotter, versioner util_xds.SnapshotVersioner) Reconciler {
+	return &reconciler{
+		hasher:      hasher,
+		cache:       cache,
+		snapshotter: snapshotter,
+		versioner:   versioner,
+	}
+}
+
+type reconciler struct {
+	hasher      envoy_cache.NodeHash
+	cache       util_xds.SnapshotCache
+	snapshotter Snapshotter
+	versioner   util_xds.SnapshotVersioner
+}
+
+func (r *reconciler) Reconcile(node *envoy_core.Node) error {
+	new, err := r.snapshotter.Snapshot(node)
+	if err != nil {
+		return err
+	}
+	id := r.hasher.ID(node)
+	old, _ := r.cache.GetSnapshot(id)
+	r.versioner.Version(new, old)
```

Comment:
What about concurrency of this method? Hasher will always return the same version. Every node has one goroutine for processing so there will be multiple goroutines operating on the same snapshot setting the version. Is it ok?

Reply:
The `new` is always a fresh object.
Anyway, `Versioner` has been changed to facilitate immutable `Snapshot`s.

Reply:
I was mostly concerned about versioning of the snapshot, but since we compare it for previous snapshot, it should be fine.